source/fab/builder.py

```diff
@@ -93,13 +94,13 @@ def entry() -> None:
 
 
 class Fab(object):
-    _extension_map: Dict[str, Union[Type[Task], Type[Command]]] = {
-        '.f90': FortranAnalyser,
-        '.F90': FortranPreProcessor,
-    }
-    _compiler_map: Dict[str, Type[Command]] = {
-        '.f90': FortranCompiler,
-    }
+    _precompile_map: List[Tuple[str, Union[Type[Task], Type[Command]]]] = [
+        (r'.*\.f90', FortranAnalyser),
+        (r'.*\.F90', FortranPreProcessor),
+    ]
+    _compile_map: List[Tuple[str, Type[Command]]] = [
+        (r'.*\.f90', FortranCompiler),
+    ]
```

Comment:
Tuples are often a convenient but ugly solution to problems. Is there a cleaner way of doing this?

Would a map of string to class be better? I imagining ordering is important and random look-up isn't.

Would a simple class with named properties make for a nicer solution? In such a case the compilation of the RegEx could be included in that object rather than having to do it later.

Reply:
Yes I think this is an obvious area for improvement - you'll see in the upcoming PR that implements "phases" that this becomes slightly uglier still (for the other map just above this).  I expect the solution is as you suggest to turn a "phase" or a "map" into an object with named properties.  But I think this might be easier after the aforementioned PR and this are both committed (what do you think?)

Reply:
In the end I had a go at making these into an object, which at least hides a bit of the ugliness

Reply:
As long as you raise an issue for it I am happy to hold off until later.